src/OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -1029,7 +985,10 @@ class SubmissionTracker : public VulkanLayerProducer::CaptureStatusListener {
   absl::flat_hash_map<VkCommandBuffer, VkDevice> command_buffer_to_device_;
 
   absl::flat_hash_map<VkCommandBuffer, CommandBufferState> command_buffer_to_state_;
-  absl::flat_hash_map<VkQueue, std::vector<QueueSubmission>> queue_to_submissions_;
+  absl::flat_hash_map<VkQueue,
```

Comment:
Have you tried just using a single priority queue? That is, one for all queues. That would simplify the code a bit more, but I'm a bit worried that a single queue might block the others in that case (e.g. long compute jobs in the compute queue).

Reply:
I haven't tried this. Unless there are compute queue jobs that takes several seconds, I would not worry too much about this. So it might be worth trying. I'll do some tests tomorrow. 

In general, I like the split here, as I actually find it more understandable. Also it may allow us to do more advanced locking in the vkQueueSubmit part.

Reply:
If my understanding is correct, then even jobs that extend over a small number of frames might cause issues, but I don't know how common this is. 

For instance, you mentioned you have seen single calls to vkQueuePresent (because of what it has to do in CompleteSubmits) of around ~33ms for all the command buffers / debug markers in a single frame. If these pile up even over only a small number of frames (>= 2) because a single job is blocking progress on the queue, this will introduce quite some hiccups. 

The ~33ms time indicates that we would probably want to distribute the work done in CompleteSubmits more rather than less. This is probably not practical and doing it in vkQueuePresent is probably what we'll keep doing. But we shouldn't make the situation worse. 

So, unless any of what I'm saying here is wrong, I'd be in favor of keeping the priority queue per Vulkan queue.

Reply:
First, I remeasured without instrumenting the "GetTimestampFunctions", in which case it was ~8ms (still pretty bad).

In general, my understanding is, that it would be still fine. We would end up in sending those events the amount of frames later, but we would also exit pretty early. There might be a problem, as we temporary might need more timer indices as before.

I would suggest to keep it as it is.

Reply:
> In general, my understanding is, that it would be still fine. We would end up in sending those events the amount of frames later, but we would also exit pretty early.

I'm worried about the case where a lot of submissions (that all finish quickly) would pile up while waiting for the topmost job to finish (which causes an early exit in CompleteSubmits). These all cause a large spike once that one job clears.

Reply:
Yeah that might happen.